cmd/kubeadm/app/util/etcd/etcd_test.go

```diff
@@ -494,3 +494,280 @@ func TestClient_GetMemberID(t *testing.T) {
+	type fields struct {
+		Endpoints      []string
+		newEtcdClient  func(endpoints []string) (etcdClient, error)
+		listMemberFunc func() (*clientv3.MemberListResponse, error)
+	}
+	tests := []struct {
+		name      string
+		fields    fields
+		memberID  uint64
+		want      bool
+		wantError bool
+	}{
+		{
+			name: "The specified member is not leader",
+			fields: fields{
+				Endpoints: []string{},
+				newEtcdClient: func(endpoints []string) (etcdClient, error) {
+					f := &fakeEtcdClient{
+						members: []*pb.Member{
+							{
+								ID:   1,
+								Name: "member1",
+								PeerURLs: []string{
+									"https://member1:2380",
+								},
+								IsLearner: false,
+							},
+						},
+					}
+					return f, nil
+				},
+			},
+			memberID: 1,
+			want:     false,
+		},
+		{
+			name: "The specified member is a leader",
+			fields: fields{
+				Endpoints: []string{},
+				newEtcdClient: func(endpoints []string) (etcdClient, error) {
+					f := &fakeEtcdClient{
+						members: []*pb.Member{
+							{
+								ID:   1,
+								Name: "member1",
+								PeerURLs: []string{
+									"https://member1:2380",
+								},
+								IsLearner: true,
+							},
+							{
+								ID:   2,
+								Name: "member2",
+								PeerURLs: []string{
+									"https://member2:2380",
+								},
+							},
+						},
+					}
+					return f, nil
+				},
+			},
+			memberID: 1,
+			want:     true,
+		},
+		{
+			name: "The specified member does not exist",
+			fields: fields{
+				Endpoints: []string{},
+				newEtcdClient: func(endpoints []string) (etcdClient, error) {
+					f := &fakeEtcdClient{
+						members: []*pb.Member{},
+					}
+					return f, nil
+				},
+			},
+			memberID: 3,
+			want:     false,
+		},
+		{
+			name: "Leader ID is empty",
+			fields: fields{
+				Endpoints: []string{},
+				newEtcdClient: func(endpoints []string) (etcdClient, error) {
+					f := &fakeEtcdClient{
+						members: []*pb.Member{
+							{
+								Name: "member2",
+								PeerURLs: []string{
+									"https://member2:2380",
+								},
+								IsLearner: true,
+							},
+						},
+					}
+					return f, nil
+				},
+			},
+			want: true,
+		},
+		{
+			name: "Listmembers return error",
+			fields: fields{
+				Endpoints: []string{},
+				newEtcdClient: func(endpoints []string) (etcdClient, error) {
+					f := &fakeEtcdClient{
+						members: []*pb.Member{
+							{
+								Name: "member2",
+								PeerURLs: []string{
+									"https://member2:2380",
+								},
+								IsLearner: true,
+							},
+						},
+					}
+					return f, nil
+				},
+				listMemberFunc: func() (*clientv3.MemberListResponse, error) {
+					return nil, errNotImplemented
+				},
+			},
+			want:      false,
+			wantError: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			c := &Client{
+				Endpoints:      tt.fields.Endpoints,
+				newEtcdClient:  tt.fields.newEtcdClient,
+				listMemberFunc: tt.fields.listMemberFunc,
```

Comment:
all of these test cases must use a fake listMemberFunc like the test `name: "Listmembers return error",`
we don't want to test listMemberFunc two times.

Reply:
Thanks, after I tried this, I found that the test code would become more redundant, like:

```go
		{
			name: "PeerURLs has multiple urls",
			fields: fields{
				Endpoints: []string{},
				newEtcdClient: func(endpoints []string) (etcdClient, error) {
					f := &fakeEtcdClient{ // The `client` is defined twice
						members: []*pb.Member{
							{
								ID:   1,
								Name: "member1",
								PeerURLs: []string{
									"https://member1:2380",
									"https://member2:2380",
								},
							},
						},
					}
					return f, nil
				},
				listMemberFunc: func() (*clientv3.MemberListResponse, error) {
					f := &fakeEtcdClient{ // 
						members: []*pb.Member{ // The `client` is defined twice
							{
								ID:   1,
								Name: "member1",
								PeerURLs: []string{
									"https://member1:2380",
									"https://member2:2380",
								},
							},
						},
					}
					resp,_ := f.MemberList(context.TODO())
					return resp, nil
				},
			},
			want: []Member{
				{
					Name:    "member1",
					PeerURL: "https://member1:2380",
				},
			},
		},
```

Is this okay?

Reply:
can it be something like this?

```
c := &Client{
	Endpoints:      tt.fields.Endpoints,
	newEtcdClient:  tt.fields.newEtcdClient,				
}
c.listMembersFunc = func( ... ) {
      return &clientv3.MemberListResponse{ members: c.newEtcdClient.members }
}
```